File: analytics/src/api/standaloneJobAnalytics.ts
import { get } from './client'

// Espelha DTOs em
// src/EfsAiHub.Core.Abstractions/Observability/StandaloneJobAnalytics.cs.
// Endpoints em
// src/EfsAiHub.Host.Api/Controllers/StandaloneJobAnalyticsController.cs.

export interface StandaloneJobByWorkflow {
  workflowId: string
  total: number
  completed: number
  failed: number
  /** Concluídos / (Concluídos + Falhos). 0..1. */
  successRate: number
  totalAttempts: number
  avgTotalMs: number
  p95TotalMs: number
}

export interface StandaloneJobOverview {
  projectId: string
  periodFrom: string
  periodTo: string
  totalJobs: number
  completed: number
  failed: number
  cancelled: number
  queued: number
  running: number
  successRate: number
  totalAttempts: number
  retryRate: number
  /** Latência queue → running em ms. */
  p95QueueMs: number
  /** Latência ponta a ponta CreatedAt → CompletedAt em ms. */
  p95TotalMs: number
  workflows: StandaloneJobByWorkflow[]
}

export interface StandaloneJobTimeseriesBucket {
  bucket: string
  created: number
  completed: number
  failed: number
  p95QueueMs: number
}

export type StandaloneJobGranularity = 'day' | 'hour'

function buildQs(projectId: string, from?: string, to?: string, extras?: Record<string, string>): string {
  const qs = new URLSearchParams()
  qs.set('projectId', projectId)
  if (from) qs.set('from', from)
  if (to) qs.set('to', to)
  if (extras) for (const [k, v] of Object.entries(extras)) qs.set(k, v)
  return `?${qs.toString()}`
}

export function getStandaloneJobSummary(
  projectId: string,
  from?: string,
  to?: string,
  init?: RequestInit,
): Promise<StandaloneJobOverview> {
  return get<StandaloneJobOverview>(
    `/analytics/standalone-jobs/summary${buildQs(projectId, from, to)}`,
    init,
  )
}

export function getStandaloneJobTimeseries(
  projectId: string,
  groupBy: StandaloneJobGranularity = 'day',
  from?: string,
  to?: string,
  init?: RequestInit,
): Promise<StandaloneJobTimeseriesBucket[]> {
  return get<StandaloneJobTimeseriesBucket[]>(
    `/analytics/standalone-jobs/timeseries${buildQs(projectId, from, to, { groupBy })}`,
    init,
  )
}

File: src/EfsAiHub.Core.Abstractions/Observability/StandaloneJobAnalytics.cs
namespace EfsAiHub.Core.Abstractions.Observability;

/// <summary>
/// Agregados de uso/saúde da fila de standalone jobs (
/// <c>aihub.background_response_jobs</c>) por projeto. Mostra como a fila tá
/// indo: throughput, taxa de sucesso, retries, jobs travados, p95 fila→running
/// e workflow→completed.
/// </summary>
public sealed class StandaloneJobOverview
{
    public required string ProjectId { get; init; }
    public DateTime PeriodFrom { get; init; }
    public DateTime PeriodTo { get; init; }

    /// <summary>Total de jobs criados no período (todos os Status).</summary>
    public long TotalJobs { get; init; }

    public long Completed { get; init; }
    public long Failed { get; init; }
    public long Cancelled { get; init; }
    public long Queued { get; init; }
    public long Running { get; init; }

    /// <summary>Concluídos / (Concluídos + Falhos). 0..1. 0 quando nenhum resolvido.</summary>
    public double SuccessRate { get; init; }

    /// <summary>Soma de <c>Attempt</c> em todos os jobs do período. Útil pra detectar workflows tóxicos.</summary>
    public long TotalAttempts { get; init; }

    /// <summary>Jobs com Attempt > 1 / total. Reflete pressão de retry.</summary>
    public double RetryRate { get; init; }

    /// <summary>Latência fila → running (CreatedAt → StartedAt) em ms. Backlog signal.</summary>
    public double P95QueueMs { get; init; }

    /// <summary>Latência ponta a ponta (CreatedAt → CompletedAt) em ms.</summary>
    public double P95TotalMs { get; init; }

    public required IReadOnlyList<StandaloneJobByWorkflow> Workflows { get; init; }
}

/// <summary>Linha agregada por <c>WorkflowId</c>. Ordenada por total DESC.</summary>
public sealed class StandaloneJobByWorkflow
{
    public required string WorkflowId { get; init; }
    public long Total { get; init; }
    public long Completed { get; init; }
    public long Failed { get; init; }
    public double SuccessRate { get; init; }
    public long TotalAttempts { get; init; }
    public double AvgTotalMs { get; init; }
    public double P95TotalMs { get; init; }
}

/// <summary>Bucket temporal de criação de jobs + sucesso/falha por bucket.</summary>
public sealed class StandaloneJobTimeseriesBucket
{
    public DateTime Bucket { get; init; }
    public long Created { get; init; }
    public long Completed { get; init; }
    public long Failed { get; init; }
    /// <summary>p95 da latência queue → running (StartedAt - CreatedAt) nesse bucket, em ms.</summary>
    public double P95QueueMs { get; init; }
}
